aplib.net-demo/Assets/Scripts/Player/Movement.cs

```diff
@@ -0,0 +1,191 @@
+            _rigidbody.AddForce(_maxSpeed * _acceleration * Time.fixedDeltaTime * _airMovementScale * _horizontalVelocity.normalized);
+
+            // Custom gravity for player while falling
+            _actualGravityScale = _rigidbody.velocity.y < 0 ? _fallingGravityScale : _normalGravityScale;
+            _rigidbody.velocity += (_actualGravityScale - 1) * _gravity * Time.fixedDeltaTime * Vector3.up;
+        }
+
+        if (isWallColliding) HandleWallCollision(wallNormal);
+        LimitVelocity(isOnSlope);
+    }
+
+    /// <summary>
+    /// Add force along the wall, to prevent sticking.
+    /// </summary>
+    /// <param name="wallNormal">The normal of the wall the player is walking against</param>
+    private void HandleWallCollision(Vector3 wallNormal) {
+        float angleMultiplier = Mathf.Clamp01(1.2f - Vector3.Angle(_horizontalVelocity, -wallNormal) / 90);
+        Vector3 directionOnWall = Vector3.ProjectOnPlane(_horizontalVelocity, wallNormal).normalized;
+        _rigidbody.AddForce(_maxSpeed * _acceleration * Time.fixedDeltaTime * angleMultiplier * directionOnWall);
+        Debug.DrawRay(transform.position, directionOnWall * 5, _physicsColor);
+    }
+
+    /// <summary>
+    ///  Limit the player's velocity to the max speed.
+    ///  Also accounts for vertical velocity when on a slope.
+    /// </summary>
+    /// <param name="isOnWalkableSlope">Is the player on a walkable slope</param>
+    private void LimitVelocity(bool isOnWalkableSlope){
+        if (isOnWalkableSlope) {
+            if (_rigidbody.velocity.magnitude > _maxSpeed) _rigidbody.velocity = _rigidbody.velocity.normalized * _maxSpeed;   
+        } else {
+            Vector3 rigidbodyHorizontalVelocity = new(_rigidbody.velocity.x, 0, _rigidbody.velocity.z);
+            if (rigidbodyHorizontalVelocity.magnitude > _maxSpeed) {
+                Vector3 limitedVelocity = rigidbodyHorizontalVelocity.normalized * _maxSpeed;
+                _rigidbody.velocity = new Vector3(limitedVelocity.x, _rigidbody.velocity.y, limitedVelocity.z);
+            }
+        }
+        
+        // Draw a ray to visualize the player's velocity and direction
+        Debug.DrawRay(transform.position, _rigidbody.velocity, _velocityColor);
+    }
+
+    /// <summary>
+    /// Check if the player is on a walkable slope.
+    /// If so, return movement direction adjusted to the slope, and the hit point.
+    /// </summary>
+    /// <param name="directionOnSlope">Movement direction projected on the slope as a plane</param>
+    /// <param name="downHit">RaycastHit of the downward slope check</param>
+    /// <returns>A boolean that signifies whether the player is on a walkable slope or not</returns>
+    private bool OnWalkableSlope(out Vector3 directionOnSlope, out RaycastHit downHit) {
+        // Shoot a ray down to check if the player is on a slope
+        if (Physics.Raycast(transform.position, Vector3.down, out downHit, _playerHeight * 0.5f + 0.3f, _groundMask)) {
```

Comment:
Where do `0.5f` and `0.3f` come from? It seems to have to do with the player height, that should definitely be defined somewhere.

Reply:
The ray starts from the player center and we want it to reach a little beyond the player's 'feet', so the ray length is 0.5 * player height + a little wiggle room. I guess we can define the wiggle room

Reply:
Defining that would be a good idea.